lib/surface/formatter/phases/final_newline.ex

```diff
@@ -1,9 +1,20 @@
 defmodule Surface.Formatter.Phases.FinalNewline do
-  @moduledoc "Add a newline after all of the nodes"
+  @moduledoc "Add a newline after all of the nodes if one was present on the original input"
 
   @behaviour Surface.Formatter.Phase
 
-  def run(nodes, _opts) do
-    nodes ++ [:newline]
+  # special case for empty heredocs
+  def run([:indent], _opts), do: []
+
+  def run(nodes, opts) do
+    suffix =
+      opts
+      |> Keyword.get(:trailing_newline_on_input, false)
+      |> final_newline()
+
+    nodes ++ suffix
   end
+
+  defp final_newline(true), do: [:newline]
+  defp final_newline(_), do: []
```

Comment:
This wouldn't materially change the code, but I'm thinking something like this would be clearer/cleaner.

```suggestion
  def run(nodes, opts) do
    if Keyword.get(opts, :trailing_newline_on_input) do
      nodes ++ [:newline]
    else
      nodes
    end
  end
```

What do you think?